XX_BubbleSort/main.cpp

```diff
@@ -0,0 +1,449 @@
+			device = physDev->createLogicalDevice(std::move(params));
+			if (!device)
+				return logFail("Failed to create a Logical Device!");
+
+			// A word about `nbl::asset::IAsset`s, whenever you see an `nbl::asset::ICPUSomething` you can be sure an `nbl::video::IGPUSomething exists, and they both inherit from `nbl::asset::ISomething`.
+			// The convention is that an `ICPU` object represents a potentially Mutable (and in the past, Serializable) recipe for creating an `IGPU` object, and later examples will show automated systems for doing that.
+			// The Assets always form a Directed Acyclic Graph and our type system enforces that property at compile time (i.e. an `IBuffer` cannot reference an `IImageView` even indirectly).
+			// Another reason for the 1:1 pairing of types is that one can use a CPU-to-GPU associative cache (asset manager has a default one) and use the pointers to the CPU objects as UUIDs.
+			// The ICPUShader is just a mutable container for source code (can be high level like HLSL needing compilation to SPIR-V or SPIR-V itself) held in an `nbl::asset::ICPUBuffer`.
+			// They can be created: from buffers of code, by compilation from some other source code, or loaded from files (next example will do that).
+			smart_refctd_ptr<nbl::asset::ICPUShader> evenPassCPUShader;
+			smart_refctd_ptr<nbl::asset::ICPUShader> oddPassCPUShader;
+			{
+				// Normally we'd use the ISystem and the IAssetManager to load shaders flexibly from (virtual) files for ease of development (syntax highlighting and Intellisense),
+				// but I want to show the full process of assembling a shader from raw source code at least once.
+				smart_refctd_ptr<nbl::asset::IShaderCompiler> compiler = make_smart_refctd_ptr<nbl::asset::CHLSLCompiler>(smart_refctd_ptr(m_system));
+
+				// A simple shader that writes out the Global Invocation Index to the position it corresponds to in the buffer
+				// Note the injection of a define from C++ to keep the workgroup size in sync.
+				// P.S. We don't have an entry point name compiler option because we expect that future compilers should support multiple entry points, so for now there must be a single entry point called "main".
+				constexpr const char* evenPass = R"===(
+						#pragma wave shader_stage(compute)
+
+						struct PushConstantData
+						{
+							uint32_t dataElementCount;
+						};
+
+						[[vk::push_constant]] PushConstantData pushConstants;
+
+						[[vk::binding(0,0)]] RWStructuredBuffer<uint32_t> buff;
+
+						[numthreads(WORKGROUP_SIZE,1,1)]
+						void main(uint32_t3 ID : SV_DispatchThreadID)
+						{
+							uint32_t N = 2 * ID.x;
+							if (N + 1 >= pushConstants.dataElementCount)
+								return;
+
+							uint32_t aux = 0;
+
+							if (buff[N] > buff[N+1]) {
+								aux = buff[N];
+								buff[N] = buff[N+1];
+								buff[N+1] = aux;
+							}
+						}
```

Comment:
if you were to refactor this into an `ext` I'd wrap the entire `main` into a struct and template on the type you want to sort, i.e.
```cpp
namespace nbl 
{
namespace hlsl 
{

struct bubble_sort
{
   template<typename Accessor, typename Comparator>
   static void __call(NBL_ARG_REF(Accessor) accessor, NBL_ARG_REF(Comparator) comp)
   {
   }
};

}
}
```

Note: because HLSL doesn't have references/pointers we use a "stateful Accessor pattern" (similar to a C++ STL iterator) see the workgroup prefix sum (ex 22) to see what we mean.